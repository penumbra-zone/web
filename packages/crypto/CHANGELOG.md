# @penumbra-zone/crypto-web

## 19.0.0

### Patch Changes

- Updated dependencies [bd43d49]
- Updated dependencies [807648a]
  - @penumbra-zone/types@18.2.0

## 18.0.0

### Patch Changes

- Updated dependencies [f5bea48]
  - @penumbra-zone/types@18.1.0

## 17.0.0

### Patch Changes

- Updated dependencies [a9ffd2d]
  - @penumbra-zone/types@18.0.0

## 16.0.1

### Patch Changes

- Updated dependencies [3477bef]
  - @penumbra-zone/types@17.0.1

## 16.0.0

### Patch Changes

- @penumbra-zone/types@17.0.0

## 15.0.0

### Patch Changes

- Updated dependencies [0233722]
  - @penumbra-zone/types@16.1.0

## 14.0.0

### Patch Changes

- @penumbra-zone/types@16.0.0

## 13.0.1

### Patch Changes

- 3aaead1: Move the "default" option in package.json exports field to the last
- Updated dependencies [3aaead1]
  - @penumbra-zone/types@15.1.1

## 13.0.0

### Patch Changes

- Updated dependencies [877fb1f]
  - @penumbra-zone/types@15.1.0

## 12.0.0

### Patch Changes

- Updated dependencies [fa798d9]
  - @penumbra-zone/types@15.0.0

## 11.0.0

### Patch Changes

- Updated dependencies [28a48d7]
  - @penumbra-zone/types@14.0.0

## 10.0.0

### Patch Changes

- Updated dependencies [43ccd96]
  - @penumbra-zone/types@13.1.0

## 9.0.0

### Patch Changes

- 3708e2c: include peer deps as dev deps
- Updated dependencies [3708e2c]
  - @penumbra-zone/types@13.0.0

## 8.0.0

### Patch Changes

- Updated dependencies
  - @penumbra-zone/types@12.0.0

## 7.0.0

### Patch Changes

- Updated dependencies
  - @penumbra-zone/types@11.0.0

## 6.0.0

### Patch Changes

- Updated dependencies [4161587]
  - @penumbra-zone/types@10.0.0

## 5.0.0

### Minor Changes

- 9b3f561: properly build esm relative paths

### Patch Changes

- Updated dependencies [9b3f561]
  - @penumbra-zone/types@9.0.0

## 4.0.0

### Major Changes

- f067fab: reconfigure all package builds

### Patch Changes

- Updated dependencies [f067fab]
  - @penumbra-zone/types@8.0.0

## 3.0.11

### Patch Changes

- @penumbra-zone/types@7.1.1

## 3.0.10

### Patch Changes

- Updated dependencies [ab9d743]
- Updated dependencies [282eabf]
- Updated dependencies [c8e8d15]
  - @penumbra-zone/types@7.1.0

## 3.0.9

### Patch Changes

- @penumbra-zone/types@7.0.1

## 3.0.8

### Patch Changes

- Updated dependencies [bb5f621]
  - @penumbra-zone/types@7.0.0

## 3.0.7

### Patch Changes

- Updated dependencies [029eebb]
- Updated dependencies [3ea1e6c]
  - @penumbra-zone/types@6.0.0

## 3.0.6

### Patch Changes

- Updated dependencies [146b48d]
- Updated dependencies [e35c6f7]
- Updated dependencies [8ccaf30]
  - @penumbra-zone/types@5.0.0

## 3.0.5

### Patch Changes

- Updated dependencies
  - @penumbra-zone/types@4.1.0

## 3.0.4

### Patch Changes

- @penumbra-zone/types@4.0.1

## 3.0.3

### Patch Changes

- Updated dependencies [6fb898a]
  - @penumbra-zone/types@4.0.0

## 3.0.2

### Patch Changes

- Updated dependencies [3148375]
  - @penumbra-zone/types@3.0.0

## 3.0.1

### Patch Changes

- @penumbra-zone/types@2.0.1

## 3.0.0

### Major Changes

- b4082b7: /src/ import path requirement removed

## 2.0.0

### Major Changes

- 929d278: barrel imports to facilitate better tree shaking

### Patch Changes

- Updated dependencies [929d278]
  - @penumbra-zone/types@2.0.0

## 1.0.1

### Patch Changes

- Updated dependencies
  - @penumbra-zone/types@1.1.0
