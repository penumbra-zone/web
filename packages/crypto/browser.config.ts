/// <reference types="vitest" />
import { defineConfig } from 'vite';

export default defineConfig({
  test: {
    include: ['**/encryption.test.ts'],
    browser: {
      name: 'chromium',
      provider: 'playwright',
      enabled: true,
      headless: true,
    },
  },
});
