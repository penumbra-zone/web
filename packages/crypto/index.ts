export * from './src/encryption';
export * from './src/sha256';
