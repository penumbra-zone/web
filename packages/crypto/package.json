{
  "name": "@penumbra-zone/crypto-web",
  "version": "14.0.0",
  "license": "(MIT OR Apache-2.0)",
  "type": "module",
  "engine": {
    "node": ">=22"
  },
  "scripts": {
    "build": "tsc --build --verbose",
    "clean": "rm -rfv dist *.tsbuildinfo package penumbra-zone-*.tgz",
    "dev:pack": "tsc-watch --onSuccess \"$npm_execpath pack\"",
    "lint": "eslint src",
    "lint:fix": "eslint src --fix",
    "lint:strict": "tsc --noEmit && eslint src --max-warnings 0",
    "test": "vitest run"
  },
  "files": [
    "dist",
    "!dist/*.test.*"
  ],
  "exports": {
    "./*": "./src/*.ts"
  },
  "publishConfig": {
    "exports": {
      "./*": {
        "types": "./dist/*.d.ts",
        "default": "./dist/*.js"
      }
    }
  },
  "dependencies": {
    "bip39": "^3.1.0",
    "crypto-js": "^4.2.0"
  },
  "devDependencies": {
    "@penumbra-zone/types": "workspace:*",
    "@types/crypto-js": "^4.2.2"
  },
  "peerDependencies": {
    "@penumbra-zone/types": "workspace:*"
  }
}
