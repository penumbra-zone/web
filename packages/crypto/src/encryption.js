"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.decrypt = exports.encrypt = exports.repeatedHash = void 0;
var sha256_1 = __importDefault(require("crypto-js/sha256"));
var aes_1 = __importDefault(require("crypto-js/aes"));
var enc_utf8_1 = __importDefault(require("crypto-js/enc-utf8"));
var STRETCH_ROUNDS = 5000;
// In cryptography, key stretching techniques are used to make brute-force attacks more difficult.
// The idea is to "stretch" the key (or in this case, the password) by applying a computationally intensive
// operation to it multiple times. This makes each attempt to guess the password more time-consuming,
// which in turn makes brute-force attacks less feasible.
var repeatedHash = function (message) {
    var hash = (0, sha256_1.default)(message);
    for (var i = 0; i < STRETCH_ROUNDS; i++) {
        hash = (0, sha256_1.default)(hash.toString());
    }
    return hash.toString();
};
exports.repeatedHash = repeatedHash;
var encrypt = function (message, key) {
    return aes_1.default.encrypt(message, key).toString();
};
exports.encrypt = encrypt;
var decrypt = function (ciphertext, key) {
    var decrypted = aes_1.default.decrypt(ciphertext, key);
    return decrypted.toString(enc_utf8_1.default);
};
exports.decrypt = decrypt;
