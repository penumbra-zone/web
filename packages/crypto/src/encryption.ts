import { Base64Str, base64ToUint8Array, uint8ArrayToBase64 } from 'penumbra-types';

// Salts & initialization vectors should be stored
export const randomBase64str = (): Base64Str => {
  const uintArr = crypto.getRandomValues(new Uint8Array(16)); // 128 bits
  return uint8ArrayToBase64(uintArr);
};

export const hashPassword = async (password: string, salt: Base64Str): Promise<JsonWebKey> => {
  const enc = new TextEncoder();
  const importedKey = await crypto.subtle.importKey(
    'raw',
    enc.encode(password),
    { name: 'PBKDF2' },
    false,
    ['deriveBits', 'deriveKey'],
  );

  const derivedKey = await crypto.subtle.deriveKey(
    {
      name: 'PBKDF2',
      salt: base64ToUint8Array(salt),
      iterations: 210_000,
      hash: 'SHA-512',
    },
    importedKey,
    { name: 'AES-GCM', length: 256 },
    true,
    ['encrypt', 'decrypt'],
  );

  return crypto.subtle.exportKey('jwk', derivedKey);
};

export const isPassword = async (
  password: string,
  salt: Base64Str,
  encryptedSeedPhrase: Base64Str,
  initializationVector: Base64Str,
): Promise<boolean> => {
  try {
    const key = await hashPassword(password, salt);
    await decrypt(encryptedSeedPhrase, initializationVector, key);
    return true;
  } catch (error) {
    return false;
  }
};

export const encrypt = async (
  message: string,
  initializationVector: Base64Str,
  jwkKey: JsonWebKey,
): Promise<Base64Str> => {
  const key = await crypto.subtle.importKey(
    'jwk',
    jwkKey,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt', 'decrypt'],
  );

  const enc = new TextEncoder();
  const buffer = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv: base64ToUint8Array(initializationVector) },
    key,
    enc.encode(message),
  );
  return uint8ArrayToBase64(new Uint8Array(buffer));
};

export const decrypt = async (
  ciphertext: Base64Str,
  initializationVector: Base64Str, // You need to provide both the same iv & key used for encryption
  jwkKey: JsonWebKey,
): Promise<string> => {
  // Import the key
  const key = await crypto.subtle.importKey(
    'jwk',
    jwkKey,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt', 'decrypt'],
  );

  const dec = new TextDecoder();
  const decrypted = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv: base64ToUint8Array(initializationVector) },
    key,
    base64ToUint8Array(ciphertext),
  );
  return dec.decode(decrypted);
};
