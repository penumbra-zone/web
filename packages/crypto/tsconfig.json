{
  "extends": "tsconfig/tsup.json",
  "include": ["src"],
  "exclude": ["node_modules", "dist"]
}
