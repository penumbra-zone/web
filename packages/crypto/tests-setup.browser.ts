import { webcrypto } from 'node:crypto';
import { vi } from 'vitest';

/**
 * JSDom (which the browser tests run in) doesn't natively support
 * `window.crypto`, so we need to polyfill it.
 */
vi.stubGlobal('crypto', webcrypto);
