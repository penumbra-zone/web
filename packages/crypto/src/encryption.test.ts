import { describe, expect, test, vi } from 'vitest';
import { decrypt, encrypt, hashPassword, isPassword, randomBase64str } from './encryption';
import { webcrypto } from 'crypto';
import { Base64StringSchema, validateSchema } from 'penumbra-types';

vi.stubGlobal('crypto', webcrypto);

describe('encryption', () => {
  const password = 's0meUs3rP@ssword';
  const seedPhrase = 'correct horse battery staple';

  test('encrypt and decrypt provide lossless round-trip', async () => {
    const salt = randomBase64str();
    const key = await hashPassword(password, salt);

    const initializationVector = randomBase64str();
    const encrypted = await encrypt(seedPhrase, initializationVector, key);
    const decrypted = await decrypt(encrypted, initializationVector, key);
    expect(decrypted).toBe(seedPhrase);
  });

  test('isPassword correctly verifies password', async () => {
    const salt = randomBase64str();
    const key = await hashPassword(password, salt);

    const initializationVector = randomBase64str();
    const encrypted = await encrypt(seedPhrase, initializationVector, key);
    const isPasswordCorrect = await isPassword(password, salt, encrypted, initializationVector);

    expect(isPasswordCorrect).toBe(true);
  });

  test('isPassword correctly rejects incorrect password', async () => {
    const salt = randomBase64str();
    const key = await hashPassword(password, salt);
    const initializationVector = randomBase64str();

    const encrypted = await encrypt(seedPhrase, initializationVector, key);
    const isPasswordCorrect = await isPassword(
      'WrongP@ssw0rd',
      salt,
      encrypted,
      initializationVector,
    );

    expect(isPasswordCorrect).toBe(false);
  });

  test('decrypt correctly rejects wrong initialization vector', async () => {
    const salt = randomBase64str();
    const key = await hashPassword(password, salt);
    const initializationVector = randomBase64str();
    const encrypted = await encrypt(seedPhrase, initializationVector, key);
    const wrongInitializationVector = randomBase64str();
    try {
      await decrypt(encrypted, wrongInitializationVector, key);
    } catch (error) {
      expect(error).toBeTruthy();
    }
  });

  test('decrypt correctly rejects wrong CryptoKey', async () => {
    const salt = randomBase64str();
    const key = await hashPassword(password, salt);
    const wrongKey = await hashPassword('wrongPassword', salt);
    const initializationVector = randomBase64str();
    const encrypted = await encrypt(seedPhrase, initializationVector, key);
    try {
      await decrypt(encrypted, initializationVector, wrongKey);
    } catch (error) {
      expect(error).toBeTruthy();
    }
  });

  test('should return a JSON serializable key', async () => {
    const salt = randomBase64str();
    const key = await hashPassword(password, salt);
    expect(JSON.parse(JSON.stringify(key))).not.toEqual({});
  });

  test('salt should be a base64 string', () => {
    const salt = randomBase64str();
    expect(() => validateSchema(Base64StringSchema, salt)).not.toThrow();
  });

  test('Hash algorithm should not have changed', async () => {
    const salt = '+VzsTs4/j3wZct7oaDhHOg==';
    const key = await hashPassword(password, salt);
    expect(key).toMatchSnapshot();
  });
});
