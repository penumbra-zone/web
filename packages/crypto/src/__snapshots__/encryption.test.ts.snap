// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encryption > Hash algorithm should not have changed 1`] = `
{
  "alg": "A256GCM",
  "ext": true,
  "k": "FnN51e8RqqZsg6lRbPk5yHIA_GSV43xYHJD6CWOlgoM",
  "key_ops": [
    "encrypt",
    "decrypt",
  ],
  "kty": "oct",
}
`;
